chapter2.adoc

```diff
@@ -1,214 +1,207 @@
-
-Memory and CSR accesses initiated by abstract commands or from the program buffer will be treated as if they are at the privilege level held in `prv` and `v`. These accesses will undergo protections of PMA, PMP, MMU, and other mechanisms, triggering traps if they violate corresponding rules. 
+The privilege level of the trigger chain is determined by the trigger enabled for the highest privilege level inside the chain. The entire trigger chain cannot be modified if the chain privilege level exceeds debug allowed privilege level.
 
 [NOTE]
-The external debugger has the capability to write to prv and v and subsequently read back the value, thus determining the maximum debug privilege level.  
-
-Additionally, the fields in dcsr are further constrained based on their sphere of action. For example, when a field is effective in machine mode, it is accessible only to debugger which is granted with machine mode privilege. The detailed accessibility is listed in the following table.
-
-.Dcsr fields accessibility against privilege granted to external debugger
-[options="header"]
-|============================================
-| Field    |  Allowed debug access privilege 
-| ebreakvs |  M/S/VS 
-| ebreakvu |  M/S/VS/VU 
-| ebreakm  |  M 
-| ebeaks   |  M/S 
-| ebreaku  |  M/S/U 
-| stepie   |  M 
-| stoptime |  M 
-| mprven   |  M 
-| nmip     |  M 
-|============================================
-
-===== Debug PC (dpc, at 0x7b1) and Debug Scratch Register (dscratch0, at 0x7b2; dscratch1, at 0x7b3)
-
-Debug PC (dpc) and Debug Scratch Register (dscratch0, dscratch1) are not restricted by prv and v fields to simplify the architecture.
+This represents a balance between usability and hardware complexity. There may be instances where the triggers are linked across different privilege levels (e.g., from S-mode to M-mode), while the external debugger may only have access with S-mode privilege. The external debugger should not modify the chain, because it could be suppressed or incorrectly match or fire in M-mode.
 
 ==== Sdtrig CSR
 
-The extension enforces access control in Debug Mode, which complicates trigger usage within Debug Mode. To mitigate these complications, certain trigger CSRs, tselect, tdata1, tdata2, tdata3, and tinfo are always permitted in Debug Mode, irrespective of the privileges granted to external debuggers. However, the remaining CSRs, tcontrol, scontext, hcontext, mcontext, and mscontext continue to adhere to the debug privileges granted.
+The CSRs tcontrol, scontext, hcontext, mcontext, and mscontext must follow access rules defined in <<dbgaccpriv, debug access privilege>>. Meanwhile, tselect, tdata1, tdata2, tdata3 are read/write when debug is allowed. The below table illustrates the accessibliby for Sdtrig CSRs.
```

Comment:
Yes, it is intentional. Because triggers are fundamental for debugger. It is not possible for M-mode SW to set breakpoint for external debugger because the action can only be set to "enter debug mode" when hart is in debug mode. Thus, we define the trigger accessibility in this way.

Reply:
I think this needs more elaboration then.  In general, debug mode accesses to M-mode CSRs will fail if mdbgen=0, so these CSRs are an exception.  If that is called out somewhere I missed it.  This will complicate CSR access units, that normally rely on address bits 9:8 to determine whether the current mode can access the CSR.  Now it will have to check the full address, so that it can treat these few trigger CSRs specially.  I think that's undesirable.  If we want to give S-mode debuggers access to triggers we should add S-mode trigger CSRs and a delegation mechanism, akin to what Smcdeleg/Ssccfg did for perf counters.  This could be useful for software as well.

> It is not possible for M-mode SW to set breakpoint for external debugger because the action can only be set to "enter debug mode" when hart is in debug mode.

In section 3.3.1 we explicitly made it possible for M-mode to access triggers with dmode=1 when mdbgen=0.  This is required for save/restore when switching between supervisor domains.  So M-mode can configure triggers for an S-mode debugger, if there's a way to invoke it.  That would save a lot of complexity.

Reply:
I think it probably worth more discussion. Regarding delegation to S-mode, the affected scope is broader than we expected because Sdsec only wants to have S-mode accessibility to trigger registers in debug mode. However, a solution akin to Smcdeleg/Ssccfg will also influence the trigger usage in non debug mode. It needs more consideration from self-hosted debugger side. 

Thanks for reminding, it is true that we relaxed dmode accessibility. But I cannot come up with a simple way to let debugger talk with M-mode code, especially when the debugger use the trigger to single step the code.

Reply:
I opened an issue #49  to continue the discussion.